src/mlpack/tests/neat_test.cpp

```diff
@@ -0,0 +1,210 @@
+/**
+ * @file neat_test.cpp
+ * @author Rahul Ganesh Prabhu
+ *
+ * Tests for NeuroEvolution of Augmenting Topologies.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#include <mlpack/core.hpp>
+
+#include <mlpack/methods/neat/neat.hpp>
+#include <mlpack/methods/neat/selection_strategies/rank_selection.hpp>
+#include <mlpack/methods/reinforcement_learning/environment/double_pole_cart.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "test_tools.hpp"
+
+using namespace mlpack;
+using namespace mlpack::rl;
+using namespace mlpack::ann;
+using namespace mlpack::neat;
+
+BOOST_AUTO_TEST_SUITE(NEATTest)
+
+/**
+ * A class that defines the XOR task.
+ *
+ * The expected output follows this scheme:
+ * |  Input  |  Output  |
+ * | {0, 0}  |    0     |
+ * | {0, 1}  |    1     |
+ * | {1, 0}  |    1     |
+ * | {1, 1}  |    0     |
+ *
+ */
+class XORTask
+{
+ public:
+  template<typename GenomeType>
+  double Evaluate(GenomeType& genome)
```

Comment:
I think this can be const as well, since we don't change the state of the given genome.

Reply:
Phased searching and cyclic evaluation both change Genome state

Reply:
So if I call ``Evaluate`` on a genome twice, same env settings, I could get different results in the testing phase?

Reply:
Let me know if I should clarify this point.

Reply:
Oops sorry, I didn't get a notification for the comment.

Yes, since cyclic networks store their activation values. I could maybe add a reset function which would clear the old activation?

Reply:
So it's just used to store temporary results; I don't think this is a problem, the only issue I see is that from a user perspective this could mean that we actually change the object. But I think, it would have a negative effect on the runtime if we make a copy of the genome first. So unless you can see a simple solution, I think we should go with the non-const solution.